### ISSUE TITLE
De-novo binder generation

### ISSUE THREAD

**Issue by kushnarang** [generatebio/chroma#14]
The [supplement](https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-023-06728-8/MediaObjects/41586_2023_6728_MOESM1_ESM.pdf) of the original Nature paper suggests that De-novo binders could be generated by "Combine (i) substructure conditioning on antigen, (ii) optional scaffold constraint on binder, and (iii) contact constraints on epitope/paratope."

From my current understanding of Chroma, I am unsure on how to use an existing antigen as the template for a de novo binder. I've been able to successfully generate re-designs on existing binders using code similar to #9 but de novo is proving more of a challenge. **In practice, what API(s) should one use to ask Chroma to design a brand new chain in complex with an existing antigen**?

As a second question, the "(iii) contact constraints" should ostensibly be implemented via the "Substructure distances conditioner" (pg 80, Supp Table 6). My understanding of this module is that it would allow the user to provide Chroma with a pre-specified binding location, akin to hotspots in RFDiffusion. I recognize all the other conditioners from the source code, but I can't find the substructure distances module. **Has that one already been implemented/open-sourced**, or does it exist as a mathematical template for now until its implemented in Python?

**Comment by vuhongai** [comment]
Hi, 
For me the opposite in fact. De novo binder design can be done directly by substructure conditioner. You just need to modify the pdb file a bit, by adding new chain (pseudo-chain) with dummy coordinate. 
Here's the [colab example](https://colab.research.google.com/drive/1O3koYjdZeIwjvHL6JDpTaCTG7PLWHNaO?usp=sharing).

**Comment by kushnarang** [comment]
Hi, thank you so much for the code example. Would you be able to share how you added a psuedo chain to the PDB, or maybe share an example? I am a little new to the PDB format.

**Comment by kushnarang** [comment]
And, have you found a way to identify particular "hotspots" for the receptor? Essentially, how to specify which receptor residues Chroma should try to diffuse the binder near?

Thank you so much!

**Comment by vuhongai** [comment]
> o share how you added a psuedo chain to the PDB, or maybe share an example? I am a little new to the PDB format.

The code is included in the colab example.

`protein = Protein(f"{input_dir}/6wrw_A.pdb", device=device) # input pdb should be clean before, in this case only the chain A of pdb 6WRW
X, C, S = protein.to_XCS()

len_binder = 100 # define length of your binder

X_new = torch.cat(
    [X,
     torch.zeros(1, len_binder, 4, 3).cuda()
    ],
    dim=1
)

C_new = torch.cat(
    [C,
     torch.full((1, len_binder), 2).cuda()
    ],
    dim=1
)

S_new = torch.cat(
    [S,
     torch.full((1, len_binder), 0).cuda()
    ],
    dim=1
)

del X,C,S

protein = Protein(X_new, C_new, S_new, device=device)
X, C, S = protein.to_XCS()`

**Comment by Samuel-gwb** [comment]
> Hi, For me the opposite in fact. De novo binder design can be done directly by substructure conditioner. You just need to modify the pdb file a bit, by adding new chain (pseudo-chain) with dummy coordinate. Here's the [colab example](https://colab.research.google.com/drive/1O3koYjdZeIwjvHL6JDpTaCTG7PLWHNaO?usp=sharing).

Thanks for sharing of the binder-designing code. 
I met one problem when using it, that the receptor sequence was changed in the final result. 
Any idea ? 
Thanks !

**Comment by Samuel-gwb** [comment]
Problem solved. Modification of the code to:
#################################
for i in range(L_complex):
    if i < L_receptor: 
        mask_aa[i] = torch.Tensor([0] * 20)
        mask_aa[i][S[0][i].item()] = 1 
    else:  
        mask_aa[i] = torch.Tensor([1] * 20)
#################################

Also ask:  how to set hot spot residues in the receptor for binder design?

**Comment by wujiewang** [comment]
Thanks for all the discussions and feedbacks! I will close this issue for now, feel free to reopen or post new issues if you feel the need.
